Architecture/TodoApp/android/core/src/main/java/com/novoda/todoapp/tasks/service/PersistedTasksService.java

```diff
@@ -254,75 +283,51 @@ private static Tasks mapFunctionToTasks(Tasks tasks, Function<SyncedData<Task>,
         };
     }
 
-    @Override
-    public Action0 deleteAllTasks() {
-        return new Action0() {
-            @Override
-            public void call() {
-                remoteDataSource.deleteAllTasks()
-                        .flatMap(deleteAllLocalTasks())
-                        .compose(asValidatedEvent(taskRelay.getValue()))
-                        .subscribe(taskRelay);
-            }
-        };
-    }
-
-    private Func1<Void, Observable<Tasks>> deleteAllLocalTasks() {
-        return new Func1<Void, Observable<Tasks>>() {
+    private Func1<Event<Tasks>, Observable<Tasks>> persistEventTasksLocally() {
+        return new Func1<Event<Tasks>, Observable<Tasks>>() {
             @Override
-            public Observable<Tasks> call(Void aVoid) {
-                return localDataSource.deleteAllTasks()
-                        .map(new Func1<Void, Tasks>() {
-                            @Override
-                            public Tasks call(Void aVoid) {
-                                return Tasks.empty();
-                            }
-                        });
+            public Observable<Tasks> call(Event<Tasks> tasksEvent) {
+                return localDataSource.saveTasks(tasksEvent.data().or(Tasks.empty()));
             }
         };
     }
 
     @Override
-    public Action0 complete(final Task originalTask) {
-        return save(originalTask.complete());
+    public void complete(final Task originalTask) {
+        save(originalTask.complete());
     }
 
     @Override
-    public Action0 activate(final Task originalTask) {
-        return save(originalTask.activate());
+    public void activate(final Task originalTask) {
+        save(originalTask.activate());
     }
 
     @Override
-    public Action0 save(final Task updatedTask) {
-        return new Action0() {
-            @Override
-            public void call() {
-                final long actionTimestamp = clock.timeInMillis();
-                remoteDataSource.saveTask(updatedTask)
-                        .compose(startAheadThenConfirmOrMarkAsError(updatedTask, actionTimestamp))
-                        .compose(updateAndPersistIfMostRecentAction())
-                        .subscribe();
-            }
-        };
+    public void save(final Task updatedTask) {
+        final long actionTimestamp = clock.timeInMillis();
+        remoteDataSource.saveTask(updatedTask)
+                .compose(startAheadThenConfirmOrMarkAsError(updatedTask, actionTimestamp))
```

Comment:
Please move `startAheadThenConfirmOrMarkAsError` after this method so it is more readable.